_posts/2022-07-18-azure-ad-authentication-cypress.md

```diff
@@ -0,0 +1,185 @@
+---
+layout: post
+font: serif
+title: Azure AD Authentication in Cypress Tests with MSAL
+date: 2022-07-18
+highlight: monokai
+author: Tim Veletta
+tags: azure ad activedirectory cypress authentication msal react
+---
+
+Its been a little over 2 years since I wrote about [Azure AD authentication in Cypress tests](2020-05-05-azure-ad-authentication-cypress.md) and as such, it is out of date. In that time, [Cypress](https://cypress.io) has gone from strength-to-strength but the big change in that time has been the [migration from the Azure Active Directory Library (ADAL) to the Microsoft Authentication Library (MSAL)](https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-migration). In this post, I will discuss how to authenticate with MSAL in your Cypress tests making reference to the change in approach from that post 2 years ago as a result of the migration.
+
+> If you'd just like to get the solution I've created a demo repo on [Github](https://github.com/MechanicalRock/cypress-msal-demo) which you can reference.
+
+The issue that originally inspired this blog post is that when a user tries to log into an app with Azure Active Directory (AD), they are redirected to a login portal to authenticate and allow the application to get a valid token. However, Cypress doesn't like this approach since it applies the same strict [same-origin](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy) policy that browsers do thus failing the test when it tries to access a different domain to your app.
```

Comment:
So when I first wrote the original blog post it flat out wasn't supported but in the block following this I mention that since v9.6.0 of Cypress, you can redirect to a new domain however its not ideal because we don't want to be testing the Azure login

Reply:
Maybe you should stress that though since it is possible to go down that road, emphasise  "its not ideal because we don't want to be testing the Azure login"

Reply:
> Since version 9.6.0, Cypress supports [multi-domain workflows](https://cypress.io/blog/2022/04/25/cypress-9-6-0-easily-test-multi-domain-workflows-with-cy-origin/) but if we use this approach, we would have to write a test that validates the Microsoft authentication flow which isn't ideal because we don't _control_ it.